lib/pengine/bundle.c

```diff
@@ -1487,6 +1488,41 @@ bundle_print_xml(pe_resource_t *rsc, const char *pre_text, long options,
     free(child_text);
 }
 
+void
+pe__output_bundle(pe_resource_t *rsc, long options, void * out)
+{
+    pcmk__output_t * _out = (pcmk__output_t *)out;
+    pe__bundle_variant_data_t *bundle_data = NULL;
+    char *child_text = NULL;
+    CRM_CHECK(rsc != NULL, return);
+
+    get_bundle_variant_data(bundle_data, rsc);
+
+    _out->begin_list(_out, "bundle", NULL, NULL);
+    _out->message(_out, "set_str_prop", "id", rsc->id);
+    _out->message(_out, "set_str_prop", "type", container_agent_str(bundle_data->agent_type));
+    _out->message(_out, "set_str_prop", "image", bundle_data->image);
+    _out->message(_out, "set_bool_prop", "unique", is_set(rsc->flags, pe_rsc_unique));
+    _out->message(_out, "set_bool_prop", "managed", is_set(rsc->flags, pe_rsc_managed));
+    _out->message(_out, "set_bool_prop", "failed", is_set(rsc->flags, pe_rsc_failed));
```

Comment:
Don't you think it would be too specific. Each datum would require it's own message that handles only it. I would propose a generic message that takes a name and an array of name-value pairs. One always have to print out name-value pairs and it's very generic and format-agnostic. 
```
{ "name_and_nvpairs", "text", pe__name_and_nvpairs_text },
{ "name_and_nvpairs", "xml", pe__name_and_nvpairs_xml }
```
The usage
```
_out->message(_out, "name_and_nvpairs", "bundle"
                  , "id", rsc->id
                  , "type", container_agent_str(bundle_data->agent_type)
                  , "image", bundle_data->image
                  , "unique", BOOL2STR(is_set(rsc->flags, pe_rsc_unique))
                  , "managed", BOOL2STR(is_set(rsc->flags, pe_rsc_managed))
                  , "failed", BOOL2STR(is_set(rsc->flags, pe_rsc_failed))
                  , NULL);
```

Reply:
All parameters will come together, but some will be filtered out depending on the options flag when printing.

Reply:
That would definitely be a useful generic method, but unfortunately it can't be used here since the same custom message will get implementations for HTML and text, which need to keep their current output, for backward compatibility. The current text output is specific to bundles.

In general, the idea behind the output model is that custom messages will be very specific, usually single-use. The reason is to allow the text output to be contextually descriptive, in natural language. Of course, there are cases where simple generic output is useful.

Reply:
Yes, but what if we give the raw name-value pairs to the name_and_nvpairs message and already inside the message we apply the filters, like 
```
switch (name) 
  case "bundle": blah; 
  case "resource": blah;
  default: 
```
We would have a set of rules, and if none of them matches, the default action would be applied. Thus we wouldn't have to export many messages, only a few and if a user wants to override the output of his datum he will simply implement another filter. I would even propose to make the name_and_nv_pairs be not a message but an pcmk__output_t function because of its very generic nature.

Reply:
I was definitely thinking of this generic approach as a pcmk__output_t method rather than a custom message, but we want those to be completely generic, i.e. no special casing of specific messages.

I don't think a switch is that much different coding-wise from the function-per-message approach. The argument parsing has to be duplicated with function-per-message, but that could be helped with a convenience function that converts name/value varargs to a hash table.

The current model favors an object-oriented approach with function pointers rather than switching. I can see arguments for both approaches, but the OO way keeps the messages speecific to each library in that library (which also avoid issues with types defined in headers that might not be available to libcrmcommon).

Reply:
> I don't think a switch is that much different coding-wise from the function-per-message approach. The argument parsing has to be duplicated with function-per-message, but that could be helped with a convenience function that converts name/value varargs to a hash table.

Yes, it of course, switch represents a hash table in pseudo code. Another advantage of using it is that it has a default action (which could also be dangerous) if there is no case-match.

I will create specific messages for each case, and leave the name_and_nvpairs in the next amend just in case. If we don't like it after the next revision, I will remove it.

Reply:
I have created those specific messages.